src/Core/Algorithms/BrainStimulator/BiotSavartSolverAlgorithm.cc

```diff
@@ -56,653 +56,556 @@ ALGORITHM_PARAMETER_DEF(BrainStimulator, VectorBField);
+	{
+		if (!preIntegration(mesh,coil))
+		{
+			return (false);
+		}
 
-					algo_->remark("number of processors:  " + boost::lexical_cast<std::string>(this->numprocessors_));
+		vmesh_->synchronize(Mesh::NODES_E | Mesh::EDGES_E);
 
-					success.resize(numprocessors_,true);
+		VMesh::Node::array_type enodes;
+		Point enode1, enode2;
 
-					//! get number of nodes for the model
-					modelSize = vmesh->num_nodes();
-					assert(modelSize > 0);
+		//! get numbder of nodes for the coil
+		coilSize_ = vcoil_->num_nodes();
 
-					try
-					{
-						matOut = new DenseMatrix(static_cast<int>(modelSize),3);
-						matOutHandle = static_cast<MatrixHandle>(matOut);
-					}
-					catch (...)
-					{
-						algo_->error("Error alocating output matrix");
-						return (false);
-					}
+		//! basic assumption
+		assert(modelSize_ > 0 && coilSize_ > 1);
 
-					return (true);
-			}
+		coilNodes_.clear();
+		coilNodes_.reserve(coilSize_);
 
-			bool PostIntegration( MatrixHandle& outdata )
+		for (VMesh::Edge::index_type i = 0; i < vcoil_->num_edges(); i++)
+		{
+			vcoil_->get_nodes(enodes,i);
+			vcoil_->get_point(enode1,enodes[0]);
+			vcoil_->get_point(enode2,enodes[1]);
+			coilNodes_.push_back(Vector(enode1));
+			coilNodes_.push_back(Vector(enode2));
+		}
+
+		//! Start the multi threaded
+		Parallel::RunTasks([this](int i) { ParallelKernel(i); }, numprocessors_);
+
+		return postIntegration(outdata);
+	}
+
+	void setIntegrationStep(double step)
+	{
+		assert(step >= 0.0);
+		extstep_ = step;
+	}
+
+	double getIntegrationStep() const
+	{
+		return extstep_;
+	}
+
+private:
+	//! integration step, will auto adapt
+	double autostep_;
+	//! integration step, externally provided
+	double extstep_;
+	//! keep nodes on the coil cached
+	std::vector<Vector> coilNodes_;
+
+	//! execute in parallel
+	void ParallelKernel(int proc_num)
+	{
+		assert(proc_num >= 0);
+
+		int cnt = 0;
+		double current = 1.0;
+		Point modelNode;
+
+		const index_type begins = (modelSize_ * proc_num) / numprocessors_;
+		const index_type ends  = (modelSize_ * (proc_num+1)) / numprocessors_;
+
+		assert( begins <= ends );
+
+		//! buffer of points used for integration
+		std::vector<Vector> integrPoints;
+		integrPoints.reserve(256);
+
+		//! keep previous step length
+		//! used for optimization purpose
+		double prevSegLen = 123456789.12345678;
+
+		//! number of integration points
+		int nips = 0;
+
+		try
+		{
+			for (index_type iM = begins; iM < ends; iM++)
 			{
-				//! check for error
-				for (size_t j=0; j<success.size(); j++)
-				{
-					if (success[j] == false) return (false);
-				}
+				vmesh_->get_node(modelNode, iM);
 
-				outdata = matOutHandle;
-
-				return (true);
-			}
-		};
+				// result
+				Vector F;
 
-
-	class PieceWiseKernel : public KernelBase
-		{
-			public:
-
-				PieceWiseKernel(const AlgorithmBase* algo, int t ) : KernelBase(algo,t)
+				for (size_t iC0 = 0, iC1 =1, iCV = 0; iC0 < coilNodes_.size(); iC0+=2, iC1+=2, iCV++)
```

Comment:
I'm all for crazy looking for loops but this one seems a little too crazy to me. Is there a better way of making this a little bit more readable?

Reply:
This was old code, it just got a white-space fix. I'm not touching the algorithms.

Reply:
oh fair enough. If you didn't write it, no touch. Cool.